### PR TITLE
Fix OSM pathfinding

### LINKED ISSUE #595: Agents get stuck in OpenStreetMapSpace example
In our example for `OpenStreetMapSpace` I've observed the following situation after about 3000 steps:

Screenshot_20220207_172811

A big share of the agents are stuck in dead ends of our map and can't seem to find their way out.
@AayushSabharwal Could it be the case that the wayfinding algorithm is broken?

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Closes #595

I have a plan to make the code a bit more sensible and easy to reason with, which should help with the bugs since there's currently a lot of scope to miss edge cases.

### CODE REVIEW

File: src/spaces/openstreetmap.jl [hunk 1]
```diff
@@ -112,7 +111,7 @@ function OpenStreetMapSpace(
     path::AbstractString;
     kwargs...
 )
-    m = graph_from_file(path; weight_type = :distance, kwargs...)
+    m = graph_from_file(path; weight_type=:distance, kwargs...)
```
Comment (Datseris) [inline src/spaces/openstreetmap.jl:114]: Why? No, please leave spaces between `=`.

Comment (AayushSabharwal) [reply src/spaces/openstreetmap.jl:114]: Thanks for pointing it out. Looks like the Julia extension is acting up, I'll just turn off the auto-formatter and do it manually.

**Comment by AayushSabharwal** [comment]
I _think_ the issue should be fixed now. I did the following inside the `agent_step!` function in zombies

> old = agent.pos
    distance_left = move_along_route!(agent, model, agent.speed * model.dt)
    npos = agent.pos
    if OSM.identical_position(old, npos, model) && !is_stationary(agent, model)
        println(agent.id, old)
    end

And it doesn't print anything for a couple thousand steps

**Review by Datseris — APPROVED** [review]
Very nice. I like a lot the decision to split the huge function into more. Actually, I think we could do it even more, but not necessary for me to submit the approval. One thing I would like a change though is to increase the code clarity of the smaller functions as I've outlined in the comments below.

**Comment by AayushSabharwal** [comment]
Sure. I'll address these in a while.

One question: Is it worth extracting `identical_position` and/or `identical_edge` to the public API? It's probably a common enough operation to check if two agents are at the same location/on the same edge. Perhaps the names can be refactored

**Comment by AayushSabharwal** [comment]
I also wonder if there's a sensible way to split this into multiple files, since it's getting a bit monolithic.

**Comment by Datseris** [comment]
Splitting into multiple files is not strictly necessary. You can have similarly useful outcome by putting appropriate headers like
```
#############################
# Section 1 with appropriate name
#############################
```
But splitting is also a good idea.

and yes for adding `OSM.same_position` and `OSM.same_road`

**Comment by AayushSabharwal** [comment]
Would `same_road` be misleading? Since there are multiple graph edges on the same road, and the function only tests for points being on the same edge.

**Comment by Datseris** [comment]
For me road == edge. For you, what is a road? I am not sure I can imagine how "multiple edges can be in the same road"...?

**Comment by AayushSabharwal** [comment]
In LightOSM, what we call "roads" are `Way`s. A `Way` is composed of multiple `Node`s connected by edges in the graph. A `Node` is not an intersection of "roads", but rather a location with a latitude and longitude (possibly with some metadata). This, for example, allows curved roads to be represented by discretization into multiple segments.

**Comment by Datseris** [comment]
But the function `nearest_road` returns the nearest edge, no?

**Comment by Datseris** [comment]
Futhermore I find it unintuitive to call Ways "roads", because the word "road" in real life isn't used like that. If I set up a google maps journey from here to another place, I would not say that I am on one single "road". On the other hand, the word "road" is also what an edge is in code. Because even a real life straight road would be several edges simply because they are mandatorily broken by intersections.

So perhaps, we should consider dropping the word road alltogether.

**Comment by AayushSabharwal** [comment]
That it does, but it's documented as returning the nearest location on a road, so I think it's okay? Since the nearest location on a road is on one of the edges that compose it.

**Comment by AayushSabharwal** [comment]
> So perhaps, we should consider dropping the word road alltogether.

Hmm. What can we use instead? `road_length` becomes `edge_length`, `nearest_road` becomes `nearest_location_on_edge`? (Off the top of my head those are the only two public functions using "road")

**Comment by Datseris** [comment]
For me it is okay to use the word road for edge. It makes sense. Let's just be clear in the documentation that throughout the OSM module, "road" refers to a directional edge on the graph. Even if in the real world the same "road" may be broken up by multiple intersections, in the OSM module all these would different "roads".

Where is the word `Way` used in either code or documentation? Is it actually used anywhere? I know that we have `OpenStreetMapPath`, but that is `path` not `way`.

**Comment by AayushSabharwal** [comment]
We haven't used `Way` anywhere, it's what LightOSM uses: https://deloittedigitalapac.github.io/LightOSM.jl/docs/types/#LightOSM.Way

**Comment by Datseris** [comment]
Go ahead and merge and tag after increasing version in Project.toml!

**Comment by AayushSabharwal** [comment]
This would be 5.1.0 right? Since `same_position` and `same_road` are new? As opposed to 5.0.1

**Pull request merged**